build: switch CHANGELOG to logchange (#4403)
kroxylicious/kroxylicious: DEV_GUIDE.md

To help simplify local testing we also have a simple composefile in `compose/kafka-compose.yaml`. See the [compose/README.md](./compose/README.md) for details about how to use the proxy deployed.

> ## Changelog Entries

Reviewer: Is Claude picking this the up?  Or do we need to add stuff to the .claude files to help him do the right thing?

Author: Added to relevant .claude file in [`2e2e8f1` (this PR)](https://github.com/kroxylicious/kroxylicious/pull/4403/commits/2e2e8f1c7b741b6e751635357a2f141dce6b10e2)

> User-facing changes must be documented by adding a YAML entry file to `changelog/unreleased/`.

Reviewer: Should we be suggesting the changelog CLI as an optional developer dependency?

Author: I personally don't think it's necessary, you can do everything through `mvn` (e.g. `mvn logchange:lint`) since you'll be installing the plugin through the root pom when building the project anyway. 

Also, manually copy/pasting/writing the changelog file/entry is simpler/faster than running `mvn logchange:add` IMO.

Reviewer: Ok, let's suck it and see.

Not every commit needs a changelog entry e.g. documentation fixes, internal build tooling changes, etc.

> **File naming:** `<issue-or-pr-number>-<short-slug>.yaml` (e.g. `1234-add-record-encryption.yaml`)

Reviewer: Observation: the workflow looks chicken-and-eggy - raise the PR to get the PR number, then reference the PR number in the changelog file.    If you are working an issue, then then issue will naturally exist first.  Maybe we just favour working from issues?

Author: Yeah, that's what I've started doing recently, previously I used the PR approach since so many other CHANGELOG entries link to a PR instead of an issue.

I think the current sentence presents both options as fair game, but we could solely mention issues if you think that's better? (this might make the dependabot changelog workflow a little confusing though since that uses the PR number)

Reviewer: I think for feature/bugs we just favour linking the changelog to issues.

Author: Changed DEV_GUIDE to only mention issues for naming scheme in: [`ddc7874` (this PR)](https://github.com/kroxylicious/kroxylicious/pull/4403/changes/ddc78742743c297cdf6003602b1870ddb53bf0e2)

**`type`** - pick the one that best describes the change, ideally following [Conventional Commits](https://www.conventionalcommits.org/):

| Type               | Use for                                                            |
|--------------------|--------------------------------------------------------------------|
| `feat`             | New features, capabilities, or configuration options               |
| `fix`              | Bug fixes                                                          |
| `security`         | Security fixes                                                     |
| `dependency_update`| Runtime dependency upgrades visible to users                       |
| `other`            | Performance improvements or user-visible refactoring               |

For example:

```yaml
# changelog/unreleased/1038-enforce-minimum-file-permissions.yaml
title: "feat(security): enforce minimum file permissions on confidential files."
type: added
issues:
    - 1038
important_notes:
    - |

>         Kroxylicious now validates that confidential files (TLS private keys, keystores, truststores, password files, KMS credential files, and AWS IRSA/Pod Identity token files) are not excessively permissive before reading them.

Reviewer: Let's make the example non-specific.

Author: Done in [`68b09ff` (this PR)](https://github.com/kroxylicious/kroxylicious/pull/4403/commits/68b09fff968dfa340818af60d43463d676adb992)

To help simplify local testing we also have a simple composefile in `compose/kafka-compose.yaml`. See the [compose/README.md](./compose/README.md) for details about how to use the proxy deployed.

## Changelog Entries

User-facing changes must be documented by adding a YAML entry file to `changelog/unreleased/`.
Not every commit needs a changelog entry e.g. documentation fixes, internal build tooling changes, etc.

**File naming:** `<issue-or-pr-number>-<short-slug>.yaml` (e.g. `1234-add-record-encryption.yaml`)

**Entry format:**
```yaml
title: "feat(runtime): add graceful shutdown with configurable drain timeout"
type: feat
authors:
  - name: Your Name
    nick: yourhandle
    url: https://github.com/yourhandle
merge_requests:    # use for PRs (only the first is rendered as a link prefix)
  - 1234
# or:
issues:            # use for issues (only the first is rendered as a link prefix)
  - 5678

> links:             # optional: link to external resources

Reviewer: (Non blocking/question) would we use this to link to the CVE record?

Author: Those links don't show up in the generated CHANGELOG so we can put whatever we want in there :shrug:.

Reviewer: We want to have a standard way to document fix CVEs.  I think if we added something like:

```
 # optional: link to external resources, CVE records etc.
```

we'd encourage uniformity.

Reviewer 2: as mentioned elsewhere I think we should favour merging sooner, and can work on the templating and completeness aspects in a subsequent PR. Ideally every piece of information in the logchange entry files would make it's way into our `CHANGELOG.md` but I don't think that should hold up this particular branch. (If we defer this it should go into an issue though).

Consistency within this branch would be good though, so if we decide to defer templating links in this branch, we should avoid mentioning those fields in the dev guide for now.

Author: Made issue in: https://github.com/kroxylicious/kroxylicious/issues/4459

To help simplify local testing we also have a simple composefile in `compose/kafka-compose.yaml`. See the [compose/README.md](./compose/README.md) for details about how to use the proxy deployed.

## Changelog Entries

> User-facing changes must be documented. Instead of editing `CHANGELOG.md` directly, add a YAML entry file to `changelog/unreleased/`. This avoids merge conflicts when multiple PRs are open.

Reviewer 3: I don't think we need to mention CHANGELOG or the reason for using separate files here. Just say what's expected.

```suggestion
User-facing changes must be documented by adding a YAML entry file to `changelog/unreleased/`. 
```

**File naming:** `<issue-or-pr-number>-<short-slug>.yaml` (e.g. `1234-add-record-encryption.yaml`)

**Entry format:**
```yaml
title: "feat(runtime): add graceful shutdown with configurable drain timeout"
type: feat
merge_requests:    # use for PRs
  - 1234
# or:
issues:            # use for issues
  - 5678
```

**Available types** - pick the one that best describes the change (ideally follow [Conventional Commits](https://www.conventionalcommits.org/)); this also controls which section the entry appears in:

> | Type         | Use for                                                                          | Section                              |

Reviewer 3: ```suggestion
| Type         | Use for                                                                          | `CHANGELOG` Section                  |
```

|--------------|----------------------------------------------------------------------------------|--------------------------------------|
| `feat`       | New features, capabilities, or configuration options (`feat`)                    | Main (under version number)          |
| `fix`        | Bug fixes                                                                        | Main (under version number)          |
| Other text   | Anything that cannot be well described using Conventional Commits                | Main (under version number)          |
| `changed`    | Behaviour changes that require user action; include migration steps in the title | "Changes, deprecations and removals" |
| `deprecated` | APIs or features being phased out; note the replacement                          | "Changes, deprecations and removals" |
| `removed`    | APIs or features that have been deleted                                          | "Changes, deprecations and removals" |

> Not every commit needs a changelog entry e.g. documentation fixes, internal build tooling changes, etc.

Reviewer 3: This would be better placed around like 630, where we're saying when we _do_ want a changelog entry.

The same issue/PR can appear in **both** sections:

- Add one YAML file with a short `type: added` summary.

  _...and..._

- A second with `type: changed` for the detailed breaking-change explanation.

For example:

```yaml
# changelog/unreleased/1038-enforce-minimum-file-permissions.yaml
title: "feat(security): enforce minimum file permissions on confidential files."
type: feat
issues:
  - 1038
```

```yaml
# changelog/unreleased/1038-file-permissions-changes.yaml
title: |
  Kroxylicious now validates that confidential files (TLS private keys, keystores, truststores, password files, KMS credential files, and AWS IRSA/Pod Identity token files) are not excessively permissive before reading them.
    * Control enforcement via `security.filePermissions.policy` in the proxy configuration:
      * `STRICT` (owner-only access, like SSH)
# ... more text ...
# ... more text ...
# ... more text ...
type: changed
issues:
  - 1038
```

For complex "Changes" entries with nested bullet points, use a YAML literal block scalar (`|`) in the `title` field.

After adding your entry file, commit it:
```shell
git add changelog/unreleased/<your-entry>.yaml

> ```

Reviewer 3: I think we can assume people will know to do this.
